File: 15.3Sum.js
var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  let i = 0;
  let result = [];

  if (nums.every((a) => a === 0)) {
    return [[0, 0, 0]];
  }

  while (i <= nums.length - 3) {
    let j = i + 1;
    let k = nums.length - 1;

    if (nums[i - 1] === nums[i]) {
      i++;
      continue;
    }

    while (j < k) {
      if (nums[j - 1] === nums[j] && j - 1 !== i) {
        j++;
        continue;
      }

      if (nums[i] + nums[j] + nums[k] < 0) {
        j++;
      } else if (nums[i] + nums[j] + nums[k] > 0) {
        k--;
      } else {
        result.push([nums[i], nums[j], nums[k]]);
        j++;
      }
    }

    i++;
  }
  return result;
};

console.log(threeSum([-1, 0, 1, 2, -1, -4]));

File: 15.3Sum/15.3Sum.go
package main

import (
	"slices"
)

func threeSum(nums []int) [][]int {
    result := [][]int{}
	slices.Sort(nums)
	// [-4 -1 -1 0 1 2]
	for i, v := range nums {
		if nums[i-1] == nums[i] {
			continue
		}

		left, right :=  i + 1, len(nums) -1

		for left < right {
			sum := v + nums[left] + nums[right]
			if sum < 0 {
				left++
			} else if sum > 0 {
				right--
			} else {
				result = append(result, []int{ i, left, right })
				left++
			}
		}
	}

	return result
}